### ISSUE TITLE
Add revoked sharing links to "Share access" button

### ISSUE THREAD

**Issue by paulproteus** [sandstorm-io/sandstorm#1476]
Context:

- In https://github.com/sandforms/sandforms/issues/39 someone noticed that a SandForms sharing link on Oasis had turned into a 403 and no one knew why.

- Upon investigation, we discovered that the token had not strictly disappeared; however, it was marked revoked, and Jack doesn't particularly remember revoking it.

Proposal:

- We should show revoked tokens in the UI. Mockup:

![sharing1](https://cloud.githubusercontent.com/assets/25457/12863990/578eabf4-cc34-11e5-9319-76e6a0afee6c.png)

![sharing2](https://cloud.githubusercontent.com/assets/25457/12863993/5790c8b2-cc34-11e5-8e4f-8e017004149d.png)

![sharing3](https://cloud.githubusercontent.com/assets/25457/12863989/578e70ee-cc34-11e5-8a4b-705bd400e9f6.png)

![sharing4](https://cloud.githubusercontent.com/assets/25457/12863992/578feb18-cc34-11e5-8679-6206c9eaa240.png)

![sharing5](https://cloud.githubusercontent.com/assets/25457/12863991/578f42da-cc34-11e5-8ab1-4837d89b9ef9.png)

- We probably don't want to show expired automatic "offer template" tokens. We might only want to show revoked tokens that have been used? Not sure it's worth the trouble to implement that, though.

Other notes:

- I ran this in bash to find out the `_id` of the token that got accidentally revoked:

```
$ js
> var Crypto = require('crypto');
undefined
> Crypto.createHash('sha256').update('NWcePQ704tnxQ20U5pDdSuDZKEnUXnjqCswU8evmlLz').digest('base64')
'JRqhzD1kn62xSA+5a/Sz+xxgCqU1Jx9pqf88XVewE8Y='
```

- I asked Jack to run this, and look at the resulting `ApiTokens` and we noticed together that the one whose `_id` matched the above output had `revoked: true` as a property.

```
ApiTokens.find({}).fetch()
```

**Comment by yakiradixon** [comment]
Showing revoked tokens is a great proposal, and I like how they're presented here.

**Comment by paulproteus** [comment]
Great!

**Comment by kentonv** [comment]
We should actually make it possible to un-revoke tokens, too.

I feel like at one point revoked tokens were displayed with a strikethrough and you could un-revoke them, but maybe I'm just thinking of the admin ipnetwork capability panel.

**Comment by zarvox** [comment]
I actually **really** dislike that property of the capability panel on principle.  Once a SturdyRef is invalidated, it seems like there's no particular reason why the app should expect it to become valid again in the future, so it should discard its reference, rather than holding on to it forever hoping that someday it will become reasonable to use again, which it won't discover until it randomly tries to restore the capability again or gets notified...some other way?

Another downside is that if there is any appreciable amount of token churn, the "revoked tokens" will outnumber the live tokens substantially, which detracts from the common workflow, which is "see what connections are still live".

Another possibly weird thing is that due to the way sharing links currently work, restoring the link would also restore access to users that might have had their access previously revoked.  It's unclear to me if this would be the expected behavior or surprising to users, and I generally wish to avoid potential security-related surprises.

I agree that we should provide some more visibility into the history of what happened to tokens, but I'm not convinced that un-revoking is actually a net win, beyond maybe the context of a limited "whoops, wrong button, please undo".  Un-revoking tokens may be super easy to implement, but that doesn't mean that it makes sense in actual workflows.  It's also arguably weird in the context of audit logging for a thing to be deleted and then undeleted - it means that you can never be sure that a link is actually "fully" deactivated, if it could be brought back to life in the future.

I'm okay with listing revoked tokens somewhere, though I'd want a way to deal with the potential for that list getting quite long, be that hiding them behind another click by default or paginating them, or something.

**Comment by kentonv** [comment]
@zarvox As a matter of principle, I think it's very important that all user actions can be undone if at all possible. So if the user accidentally revokes the wrong link, they should be able to restore it.

**Comment by zarvox** [comment]
Reasonable.  That can be implemented in a manner consistent with non-undoable revocation under the hood by delaying the actual revocation until the user's undo interval has passed.  Something akin to how Gmail's Undo Send works.

**Comment by paulproteus** [comment]
Hi Drew!

> I'm okay with listing revoked tokens somewhere, though I'd want a way to
deal with the potential for that list getting quite long, be that hiding
them behind another click by default or paginating them, or something.

Glad to hear it. Did you look at my mockup of a proposal for hiding the
revoked tokens behind a click? I'm curious to hear your opinion on that.

>

**Comment by zarvox** [comment]
@paulproteus Yeah, +1 on the "discoverable-but-not-in-the-list-by-default" mockup here.  I'm mostly opposed to burdening users with forming a mental model for revokable-but-later-restorable links, and how that impacts understanding who has access to things.  Misclicks aside, if you're revoking a link, probably the link got leaked and you actually want it to remain dead.

**Comment by paulproteus** [comment]
Yeah. I think I'm gently +1 on "revoking a link can't be undone", but then
I want to let people undo it, so then I want a GMail "Undo Send"-type UI,
where clicking it gives the user about 10 seconds to undo it.

**Comment by kentonv** [comment]
Note that the exact case where this came up would *not* have been solved by a 10-second timeout. Jack went looking for the revoked link some time later and presumably would have liked to un-revoke it since people were trying to use it.

Speaking as a user, I really hate gmail's 10-second timeout, and wish I could increase it to at least 60 seconds (with corresponding increase in outgoing email latency). Gmail has no choice but to implement some sort of timeout since they can't take back a message once it is sent. I don't think we should copy them if we don't have the same constraints.

**Comment by paulproteus** [comment]
I think Jack clicked the (X) by mistake, and so IMHO he would have reacted
to a thing that caused a visual jump in page layout that said he had 10
seconds to undo.

We can discuss more later, but that's why I personally do think it would
have helped Jack. I guess we can see what he thinks if he comments on this.
